packages/flutter/lib/src/material/material_state.dart

```diff
@@ -308,27 +274,78 @@ class _EnabledAndDisabledMouseCursor extends MaterialStateMouseCursor {
   String get debugDescription => 'MaterialStateMouseCursor($name)';
 }
 
-/// Interface for classes that can return a value of type `T` based on a set of
-/// [MaterialState]s.
+/// Interface for classes that [resolve] to a value of type `T` based
+/// on a widget's interactive "state", which is defined as a set
+/// of [MaterialState]s.
+///
+/// Material state properties represent values that depend on a widget's material
+/// "state".  The state is encoded as a set of [MaterialState] values, like
+/// [MaterialState.focused], [MaterialState.hovered], [MaterialState.pressed].  For
+/// example the [InkWell.overlayColor] defines the color that fills the ink well
+/// when it's pressed (the "splash color"), focused, or hovered. The [InkWell]
+/// uses the overlay color's [resolve] method to compute the color for the
+/// ink well's current state.
+///
+/// [ButtonStyle], which is used to configure the appearance of
+/// buttons like [TextButton], [ElevatedButton], and [OutlinedButton],
+/// has many material state properties.  The button widgets keep track
+/// of their current material state and [resolve] the button style's
+/// material state properties when their value is needed.
+///
+/// {@tool snippet}
```

Comment:
Would it be possible/useful to make this one of those interactive dartpad examples? The users could hover over the button and see the color change for themselves.

Reply:
Additionally, it might be useful to show hovered or focused: blue, pressed: green, default: red in this interactive example so users can see the different states in their browser.

Reply:
I've made the examples interactive. 

I didn't change this example as you've suggested because I wanted developers to see the `states.any(interactiveStates.contains)` idiom.